File: db4on/reference/db4o-reference-chapters/src/com/db4odoc/f1/evaluations/EvaluationExample.cs
using System.IO;
using com.db4o;
using com.db4o.query;

namespace com.db4odoc.f1.evaluations
{
	public class EvaluationExample 
	{
		public readonly static string YapFileName = "formula1.yap";

		public static void Main(string[] args)
		{
			File.Delete(YapFileName);
			ObjectContainer db = Db4o.OpenFile(YapFileName);
			try
			{
				StoreCars(db);
				QueryWithEvaluation(db);
			}
			finally
			{
				db.Close();
			}
		}
		// end Main

		public static void StoreCars(ObjectContainer db)
		{
			Pilot pilot1 = new Pilot("Michael Schumacher", 100);
			Car car1 = new Car("Ferrari");
			car1.Pilot = pilot1;
			car1.Snapshot();
			db.Set(car1);
			Pilot pilot2 = new Pilot("Rubens Barrichello", 99);
			Car car2 = new Car("BMW");
			car2.Pilot = pilot2;
			car2.Snapshot();
			car2.Snapshot();
			db.Set(car2);
		}
		// end StoreCars

		public static void QueryWithEvaluation(ObjectContainer db)
		{
			Query query = db.Query();
			query.Constrain(typeof (Car));
			query.Constrain(new EvenHistoryEvaluation());
			ObjectSet result = query.Execute();
			ListResult(result);
		}
		// end QueryWithEvaluation

		public static void ListResult(ObjectSet result)
		{
			System.Console.WriteLine(result.Count);
			foreach (object item in result)
			{
				System.Console.WriteLine(item);
			}
		}
		// end ListResult
	}
}

File: db4on/reference/vb/db4o-reference-chapters/src/com/db4odoc/f1/evaluations/EvaluationExample.vb
Imports System.IO
Imports com.db4o.query
Imports com.db4o

Namespace com.db4odoc.f1.evaluations
    Public Class EvaluationExample
        Public Shared ReadOnly YapFileName As String = "formula1.yap"

        Public Shared Sub Main(ByVal args As String())
            File.Delete(YapFileName)
            Dim db As Global.com.db4o.ObjectContainer = Db4oFactory.OpenFile(YapFileName)
            Try
                StoreCars(db)
                QueryWithEvaluation(db)
            Finally
                db.Close()
            End Try
        End Sub
        ' end Main

        Public Shared Sub StoreCars(ByVal db As ObjectContainer)
            Dim pilot1 As evaluations.Pilot = New evaluations.Pilot("Michael Schumacher", 100)
            Dim car1 As evaluations.Car = New evaluations.Car("Ferrari")
            car1.Pilot = pilot1
            car1.Snapshot()
            db.[Set](car1)
            Dim pilot2 As evaluations.Pilot = New evaluations.Pilot("Rubens Barrichello", 99)
            Dim car2 As evaluations.Car = New evaluations.Car("BMW")
            car2.Pilot = pilot2
            car2.Snapshot()
            car2.Snapshot()
            db.[Set](car2)
        End Sub
        ' end StoreCars

        Public Shared Sub QueryWithEvaluation(ByVal db As ObjectContainer)
            Dim query As Query = db.Query()
            query.Constrain(GetType(evaluations.Car))
            query.Constrain(New EvenHistoryEvaluation())
            Dim result As ObjectSet = query.Execute()
            Util.ListResult(result)
        End Sub
        ' end QueryWithEvaluation
    End Class
End Namespace
